pkg/providers/tf/provider.go

```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		return models.ServiceInstanceDetails{}, err
+	}
+
+	return models.ServiceInstanceDetails{
+		OperationId:   tfId,
+		OperationType: models.ProvisionOperationType,
+	}, nil
+}
+
+// Bind creates a new backing Terraform job and executes it, waiting on the result.
+func (provider *terraformProvider) Bind(ctx context.Context, bindContext *varcontext.VarContext) (map[string]interface{}, error) {
+	provider.logger.Info("bind", lager.Data{
+		"context": bindContext.ToMap(),
+	})
+
+	tfId, err := provider.create(ctx, bindContext, provider.serviceDefinition.BindSettings)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := provider.jobRunner.Wait(ctx, tfId); err != nil {
+		return nil, err
+	}
+
+	return provider.jobRunner.Outputs(ctx, tfId, wrapper.DefaultInstanceName)
+}
+
+func (provider *terraformProvider) create(ctx context.Context, vars *varcontext.VarContext, action TfServiceDefinitionV1Action) (string, error) {
+	tfId := vars.GetString("tf_id")
+	if err := vars.Error(); err != nil {
+		return "", err
+	}
+
+	workspace, err := wrapper.NewWorkspace(vars.ToMap(), action.Template)
+	if err != nil {
+		return tfId, err
+	}
+
+	if err := provider.jobRunner.StageJob(ctx, tfId, workspace); err != nil {
+		return tfId, err
+	}
+
+	return tfId, provider.jobRunner.Create(ctx, tfId)
+}
+
+// BuildInstanceCredentials unions the OtherDetails maps of the instance and bind records.
+func (provider *terraformProvider) BuildInstanceCredentials(ctx context.Context, bindRecord models.ServiceBindingCredentials, instanceRecord models.ServiceInstanceDetails) (map[string]interface{}, error) {
+	vc, err := varcontext.Builder().
+		MergeJsonObject(json.RawMessage(bindRecord.OtherDetails)).
+		MergeJsonObject(json.RawMessage(instanceRecord.OtherDetails)).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+
+	return vc.ToMap(), nil
+}
+
+// Unbind performs a terraform destroy on the binding.
+func (provider *terraformProvider) Unbind(ctx context.Context, instanceRecord models.ServiceInstanceDetails, bindRecord models.ServiceBindingCredentials) error {
+	tfId := generateTfId(instanceRecord.ID, bindRecord.BindingId)
+	provider.logger.Info("unbind", lager.Data{
+		"instance": instanceRecord.ID,
+		"binding":  bindRecord.ID,
+		"tfId":     tfId,
+	})
+
+	if err := provider.jobRunner.Destroy(ctx, tfId); err != nil {
+		return err
+	}
+
+	return provider.jobRunner.Wait(ctx, tfId)
+}
+
+// Deprovision performs a terraform destroy on the instance.
+func (provider *terraformProvider) Deprovision(ctx context.Context, instance models.ServiceInstanceDetails, details brokerapi.DeprovisionDetails) (operationId *string, err error) {
+	provider.logger.Info("terraform-deprovision", lager.Data{
+		"instance": instance.ID,
+	})
+
+	tfId := generateTfId(instance.ID, "")
+	if err := provider.jobRunner.Destroy(ctx, tfId); err != nil {
+		return nil, err
+	}
+
+	return &tfId, nil
```

Comment:
Why is there different behavior between the Unbind waiting and checking the result of the job runner's Destroy, and the Deprovision not waiting? 

Since both are terraform applys, it seems equally likely that the Deprovision could fail and leave behind infrastructure.